Rich Text: Indicate which text will be turned into a link

noisysocks opened the pull request:
Fixes #6156. Supersedes https://github.com/WordPress/gutenberg/pull/4909.

When inserting a link, the text selection disappears when the focus changes into the `URLInput` text field. This makes it hard to tell which text will be turned into a link.

The Classic Editor solves this by inserting a placeholder `<a>` element, which is the approach that I'm borrowing here.

The placeholder `<a>` is removed when the selection changes or when the block is deselected. We set `data-mce-bogus` on the placeholder `<a>` so that [it doesn't ever get serialised into post content](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/rich-text/format.js#L60).

![links](https://user-images.githubusercontent.com/612155/43940982-8f22f7b0-9cb5-11e8-95c8-c2e5c961b9cd.gif)

**To test:**

I've included E2E tests for most basic link interactions, so look for the ✅!

It's worthwhile manually testing that no undo levels are created when placeholder links are created and destroyed. To do this:

1. Add a `console.log( 'undo level created' )` to `RichText::onCreateUndoLevel()`
1. Select some text
2. Press the _Link_ button in the formatting toolbar
3. Verify that the selected text looks like a link and that no undo level was created
4. Click on something else
3. Verify that there's no longer a link and that no undo level was created
4. Create a link for real this time
5. Verify that a link appears and that an undo level was created

noisysocks committed:
Rich Text: Indicate which text will be turned into a link

When inserting a link, the text selection disappears when the focus
changes into the URLInput text field. This makes it hard to tell which
text will be turned into a link.

The Classic Editor solves this by inserting a placeholder <a> element,
which is the approach that we borrow here.

noisysocks committed:
Add E2E tests for creating, editing and removing links

talldan commented on test/e2e/specs/managing-links.test.js:
```diff
@@ -2,15 +2,178 @@
 		await newPost();
 	} );
 
+	it( 'Creating a link by selecting text and clicking the Link UI', async () => {
```
I think this should be something like this, so that it makes a nice sentence:
```
describe( 'Creating a link by selecting text and clicking the Link UI', () => {
  it( 'creates the link placeholder', async () => {
```

noisysocks replied:
I've rejigged the `describe()` and `it()` descriptions in e5ef740b0 to read like English sentences 🇬🇧

talldan reviewed (dismissed):
Works well, and code looks good, though I'm not particularly familiar with the workings of RichText. This was a good chance to learn a few things.

I think the test descriptions could be improved, happy with the rest.

noisysocks committed:
Use an `if` instead of a `switch` in getFormatValue()

A `switch` is overkill when there's only one format value that acts
differently.

noisysocks committed:
Rename Link E2E tests

Renames managing-links → links and makes the test descriptions read like
English sentences.

noisysocks commented on packages/editor/src/components/rich-text/index.js:
```diff
@@ -58,15 +59,25 @@ const { Node, getSelection } = window;
 		case 'link' : {
-			const anchor = find( parents, ( node ) => node.nodeName.toLowerCase() === 'a' );
-			return !! anchor ? { value: anchor.getAttribute( 'href' ) || '', target: anchor.getAttribute( 'target' ) || '', node: anchor } : {};
+			const anchor = find( parents, ( node ) => node.nodeName === 'A' );
```
Yeah, it's probably not useful. Removed in 7d2ffcea4 👌

noisysocks commented:
> I'm not particularly familiar with the workings of RichText. This was a good chance to learn a few things.

Ah, me too! 😆 

Thanks for the review. I'd like some 👀 from @iseulde or @azaozz before continuing since they're the in-house TinyMCE experts.

azaozz commented:
Played with this a bit more. Works well as far as I can tell. Was a bit concerned that we do `find()` on the nodes passed from TinyMCE's `nodechange` that may not include the end node in a range in some cases (as far as I remember), but wasn't able to trigger that.

Noticed some general inconsistencies with adding a link though. Not sure if any is related to this pr, but sometimes undo doesn't undo just the action of adding a link. It removes/reverts most of the typing together with the link as one step, and redo add everything back in one step.

Testing in the Classic block, it behaves quite inconsistently, especially when the (old) link modal is used (by clicking the cogwheel there). That will probably need another issue if one doesn't exist already.

azaozz commented:
Uh, one more: when clicking in the middle of a word (not selecting it) and clicking the Add link button in the Classic editor the word is selected automatically. Here the word is not selected. That results in inserting a link where the linked text is the whole URL. So we end up with the word split by an linked URL.

Thinking this should work like in the Classic editor and select the word where the caret is as soon as Add link is clicked. It works that way when setting to "bold" for example.

noisysocks commented:
Thanks @azaozz! Since those issues you've identified exist in `master` let's ticket and address them separately. 

> Noticed some general inconsistencies with adding a link though. Not sure if any is related to this pr, but sometimes undo doesn't undo just the action of adding a link. It removes/reverts most of the typing together with the link as one step, and redo add everything back in one step.

Could you create an issue for this? I'm unable to reproduce it in my environment.

> Thinking this should work like in the Classic editor and select the word where the caret is as soon as Add link is clicked. It works that way when setting to "bold" for example.

Agree with this. Let's tackle it in https://github.com/WordPress/gutenberg/issues/9183.

noisysocks commented:
@talldan @azaozz: Can one of you approve this to make GitHub happy? 🙂

aduth commented on test/e2e/specs/links.test.js:
```diff
@@ -0,0 +1,230 @@
+		expect( await getEditedPostContent() ).toMatchSnapshot();
+	} );
+
+	const setFixedToolbar = async ( b ) => {
```
`b`? What is `b`?

Verbose good. Cryptic bad.
